python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,502 @@
+def _(value: Decimal) -> Literal[Decimal]:
+    return DecimalLiteral(value)
+
+
+class AboveMax(Literal[None], Singleton):
+    def __init__(self):
+        pass
+
+    def value(self):
+        raise ValueError("AboveMax has no value")
+
+    def to(self, type_var):
+        raise TypeError("Cannot change the type of AboveMax")
+
+    def __repr__(self):
+        return "AboveMax()"
+
+    def __str__(self):
+        return "AboveMax"
+
+
+class BelowMin(Literal[None], Singleton):
+    def __init__(self):
+        pass
+
+    def value(self):
+        raise ValueError("BelowMin has no value")
+
+    def to(self, type_var):
+        raise TypeError("Cannot change the type of BelowMin")
+
+    def __repr__(self):
+        return "BelowMin()"
+
+    def __str__(self):
+        return "BelowMin"
+
+
+class BooleanLiteral(Literal[bool]):
+    def __init__(self, value):
+        self._value = value
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(BooleanType)
+    def _(self, type_var):
+        return self
+
+
+class IntegerLiteral(Literal[int]):
+    def __init__(self, value):
+        self._value = value
+
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(IntegerType)
+    def _(self, type_var: IntegerType) -> "IntegerLiteral":
+        return self
+
+    @to.register(LongType)
+    def _(self, type_var: LongType) -> "LongLiteral":
+        return LongLiteral(self.value)
+
+    @to.register(FloatType)
+    def _(self, type_var: FloatType) -> "FloatLiteral":
+        return FloatLiteral(float(self.value))
+
+    @to.register(DoubleType)
+    def _(self, type_var: DoubleType) -> "DoubleLiteral":
+        return DoubleLiteral(self.value)
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return DateLiteral(self.value)
+
+    @to.register(DecimalType)
+    def _(self, type_var: DecimalType) -> "DecimalLiteral":
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value))
+        else:
+            return DecimalLiteral(
+                Decimal(self.value).quantize(
+                    Decimal("." + "".join(["0" for _ in range(1, type_var.scale)]) + "1"), rounding=ROUND_HALF_UP
+                )
+            )
+
+
+class LongLiteral(Literal[int]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(LongType)
+    def _(self, type_var: LongType) -> "LongLiteral":
+        return self
+
+    @to.register(IntegerType)
+    def _(self, type_var: IntegerType) -> Union[AboveMax, BelowMin, IntegerLiteral]:
+        if IntegerType.max < self.value:
+            return AboveMax()
+        elif IntegerType.min > self.value:
+            return BelowMin()
+        return IntegerLiteral(self.value)
+
+    @to.register(FloatType)
+    def _(self, type_var: FloatType) -> "FloatLiteral":
+        return FloatLiteral(self.value)
```

Comment:
This is missing the conversion to `float`. Is there a way to add type checking when constructing literals like `FloatLiteral`?

Reply:
added `float(value)` here. From what I can tell, the parameter of `FloatLiteral` constructor is set to `float` but one can still pass with int or other type, but its `_value32` with round trip to struct seem to be float

Reply:
> Is there a way to add type checking when constructing literals

The [typeguard](https://github.com/agronholm/typeguard) library is a common alternative to explicitly checking the type with if statements, the `@typechecked` decorator in particular is what I've seen used a lot.

Reply:
I just mean that we want type checking to help us catch places where we're passing the wrong type into classes like `FloatLiteral`. I'll look more into this locally.

Reply:
Oh I see, I would have expected mypy to catch this when the tests ran since it should be able to deduce that the `T` in `value: T` is an int. It's possible that this might be a gap in mypy where it's not factoring in that LongLiteral inherits from `Literal[int]`.

Typically though mypy will raise an error during tests along the lines of `Argument 1 to "__init__" has incompatible type "int"; expected "float"`.